### ISSUE TITLE
"implementation detail" message has newline in it.

### ISSUE THREAD

**Issue by coke** [rakudo/rakudo#3723]
Not sure if it's intentional, but the message is short enough it doesn't need the newline.

## The Problem

```
> Attribute.^methods[0]
The 'ForeignCode' class is a Rakudo-specific
implementation detail and has no serviceable parts inside
```
## Expected Behavior

```
> Attribute.^methods[0]
The 'ForeignCode' class is a Rakudo-specific implementation detail and has no serviceable parts inside
```

**Comment by coke** [comment]
Also maybe add trailing period?

**Comment by tbrowder** [comment]
You said you usually have a terminal available to show 120 chars.  I don't. I still owe you some references on recommended text line lengths for the USAGE line wrap issue.

**Comment by lizmat** [comment]
The message would not be short enough for something like `Rakudo::SlippyOperator`  :-)

**Comment by lizmat** [comment]
More generally: some error messages in Rakudo have a newline in them, some don't even if they are too long.  Perhaps we should remove all newlines from error messages, and adapt the output to match 72 characters with some word wrapping logic.

**Comment by tbrowder** [comment]
FYI, the GNU Coding Standards recommends a maximum of 79 characters in source code.

**Comment by lizmat** [comment]
Some research shows that some `method message` return values are actually a list of possible reasons. e.g. `X::Supply::New` and `X::NYI`.  But there are plenty of other cases where the message is clearly longer than 79 characters, and no newline is provided.  So maybe we should turn this into a nice beginners task:

- look up all `method message` in core
- if they're one large text that can be auto-wrapped, remove newlines
- if they're a text that needs newlines, make sure that they are there at the right place
- make a simple word wrapper for outputting error messages

**Comment by jnthn** [comment]
Yes, putting the burden of wrapping at the point of output feels more better than spreading it all over the `message` implementations. Though (I think this is the plan, but just in case) I'd make sure the word wrapper is kept as an internal detail; doing a proper one is an entire Unicode Annex worth of fun...

**Comment by lizmat** [comment]
Yeah, I was just thinking on using `words` and use the number of chars to determine wrapping points.

**Comment by lizmat** [comment]
Hmm... I'm unsure where `.message` actually gets called.

**Comment by lizmat** [comment]
I guess in `Exception.print_exception`

**Comment by lizmat** [comment]
Made the implementation detail message word wrap, as originally requested.

With regards to automatically word wrapping all error messages: it feels that this should be part of a more general error message overhaul which is going to be needed to handle localization of error messages properly.

Until then, closing this.
